\documentclass[10pt,letterpaper]{article}
\usepackage[top=0.85in,left=2.75in,footskip=0.75in]{geometry}

% amsmath and amssymb packages, useful for mathematical formulas and symbols
\usepackage{amsmath,amssymb}

% Use adjustwidth environment to exceed column width (see example table in text)
\usepackage{changepage}

% Use Unicode characters when possible
\usepackage[utf8x]{inputenc}

% textcomp package and marvosym package for additional characters
\usepackage{textcomp,marvosym}

% cite package, to clean up citations in the main text. Do not remove.
%\usepackage{cite}

\usepackage[sort&compress,numbers]{natbib}
%\bibliographystyle{unsrtnat}
\bibliographystyle{plos2015}

% Use nameref to cite supporting information files (see Supporting Information section for more info)
\usepackage{nameref,hyperref}

\hypersetup{
    colorlinks,
    linkcolor={red!50!black},
    citecolor={blue!50!black},
    urlcolor={blue!80!black}
}

% line numbers
\usepackage[right]{lineno}

% ligatures disabled
\usepackage{microtype}
\DisableLigatures[f]{encoding = *, family = * }

% color can be used to apply background shading to table cells only
\usepackage[table]{xcolor}

% array package and thick rules for tables
\usepackage{array}

% create "+" rule type for thick vertical lines
\newcolumntype{+}{!{\vrule width 2pt}}

% create \thickcline for thick horizontal lines of variable length
\newlength\savedwidth
\newcommand\thickcline[1]{%
  \noalign{\global\savedwidth\arrayrulewidth\global\arrayrulewidth 2pt}%
  \cline{#1}%
  \noalign{\vskip\arrayrulewidth}%
  \noalign{\global\arrayrulewidth\savedwidth}%
}

% \thickhline command for thick horizontal lines that span the table
\newcommand\thickhline{\noalign{\global\savedwidth\arrayrulewidth\global\arrayrulewidth 2pt}%
\hline
\noalign{\global\arrayrulewidth\savedwidth}}


% Remove comment for double spacing
%\usepackage{setspace} 
%\doublespacing

% Text layout
\raggedright
\setlength{\parindent}{0.5cm}
\textwidth 5.25in 
\textheight 8.75in

% Bold the 'Figure #' in the caption and separate it from the title/caption with a period
% Captions will be left justified
\usepackage[aboveskip=1pt,labelfont=bf,labelsep=period,justification=raggedright,singlelinecheck=off]{caption}
\renewcommand{\figurename}{Fig}

% Remove brackets from numbering in List of References
\makeatletter
\renewcommand{\@biblabel}[1]{\quad#1.}
\makeatother

% Header and Footer with logo
\usepackage{lastpage,fancyhdr,graphicx}
\usepackage{epstopdf}
%\pagestyle{myheadings}
\pagestyle{fancy}
\fancyhf{}
%\setlength{\headheight}{27.023pt}
%\lhead{\includegraphics[width=2.0in]{PLOS-submission.eps}}
\rfoot{\thepage/\pageref{LastPage}}
\renewcommand{\headrulewidth}{0pt}
\renewcommand{\footrule}{\hrule height 2pt \vspace{2mm}}
\fancyheadoffset[L]{2.25in}
\fancyfootoffset[L]{2.25in}
\lfoot{\today}

%% END MACROS SECTION

\widowpenalty10000
\clubpenalty10000

\begin{document}
\vspace*{0.2in}

% Title must be 250 characters or less.
\begin{flushleft}
{\Large
\textbf\newline{Tximeta: reference sequence checksums for provenance identification in RNA-seq}
}
\newline
\\
Michael I. Love\textsuperscript{1,2*},
Charlotte Soneson\textsuperscript{3,4},
Peter F. Hickey\textsuperscript{5,6},
Lisa K. Johnson\textsuperscript{7},
N. Tessa Pierce\textsuperscript{7},
Lori Shepherd\textsuperscript{8},
Martin Morgan\textsuperscript{8},
Rob Patro\textsuperscript{9}
\\
\bigskip
\textbf{1} Department of Biostatistics, University of North Carolina-Chapel Hill, Chapel Hill, NC, USA
\\
\textbf{2} Department of Genetics, University of North Carolina-Chapel Hill, Chapel Hill, NC, USA
\\
\textbf{3} Friedrich Miescher Institute for Biomedical Research, Basel, Switzerland
\\
\textbf{4} SIB Swiss Institute of Bioinformatics, Basel, Switzerland
\\
\textbf{5} Epigenetics and Development Division, The Walter and Eliza Hall Institute of Medical Research, 1G Royal Parade, Parkville VIC 3052, Australia
\\
\textbf{6} The Department of Medical Biology, University of Melbourne, Parkville VIC 3010, Australia
\\
\textbf{7} Department of Population Health and Reproduction, University of California, Davis, Davis, California, USA
\\
\textbf{8} Roswell Park Comprehensive Cancer Center, Buffalo, NY, USA
\\
\textbf{9} Department of Computer Science, University of Maryland, College Park, MD, USA
\\
\bigskip

% Use the asterisk to denote corresponding authorship and provide email address in note below.
* michaelisaiahlove@gmail.com

\end{flushleft}

% Please keep the abstract below 300 words
\section*{Abstract}

Correct annotation metadata is critical for reproducible and accurate
RNA-seq analysis. When files are shared publicly or among
collaborators with incorrect or missing annotation metadata, it
becomes difficult or impossible to reproduce bioinformatic analyses
from raw data. It also makes it more difficult to locate the
transcriptomic features, such as transcripts or genes, in their proper
genomic context, which is necessary for overlapping expression data
with other datasets. We provide a solution in the form of an
R/Bioconductor package tximeta that performs numerous annotation and
metadata gathering tasks automatically on behalf of users during the
import of transcript quantification files. The correct reference
transcriptome is identified via a hashed checksum stored in the
quantification output, and key transcript databases are downloaded and
cached locally. The computational paradigm of automatically adding
annotation metadata based on reference sequence checksums can greatly
facilitate genomic workflows, by helping to reduce overhead during
bioinformatic analyses, preventing costly bioinformatic mistakes, and
promoting computational reproducibility.
The tximeta package is
available at \url{https://bioconductor.org/packages/tximeta}.

\linenumbers

\input{main}

\section*{Acknowledgments}

The authors thank the following individuals for useful discussions 
in the development of tximeta: Vince Carey, Paul Flicek, Joel Parker,
Oliver Hoffmann, Stephen Turner, Shannan Ho Sui, Thomas Keane, Andy
Yates, Reece Hart, Matthew Laird, Terence Murphy, Nathan Sheffield.
The authors also thank Reid Brennan, and C. Titus Brown, and Andrew
Whitehead for allowing use of the killifish transcriptome dataset in
the \textit{de novo} transcriptome example.

\section*{Funding}

M.I.L. is supported by R01 HG009937, R01 MH118349, P01 CA142538 and
P30 ES010126.
NTP is supported by NSF PRFB 1711984.
L.S. and M.M. are supported by U41 HG004059.
R.P. is supported by R01 HG009937, BIO-1564917, CCF-1750472, and
CNS-1763680. 

\nolinenumbers

\bibliography{refs}

\end{document}
